overlord/configstate/configcore/services.go

```diff
@@ -231,6 +236,85 @@ func handleServiceDisableConfiguration(_ sysconfig.Device, tr config.ConfGetter,
 			}
 		}
 	}
+	// configure ssh ports
+	if err := handleNetworkConfigSSHPort(dev, tr, opts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateServiceConfiguration(tr config.ConfGetter) error {
+	// validate the ssh port setting
+	output, err := coreCfg(tr, sshPortOpt)
+	if err != nil {
+		return err
+	}
+	if output == "" {
+		return nil
+	}
+	port, err := strconv.Atoi(output)
+	if err != nil {
+		return err
+	}
+	if port > 65535 || port < 1 {
+		return fmt.Errorf("cannot use port %v: must be in the range 1-65535", port)
+	}
+	return nil
+}
+
+func handleNetworkConfigSSHPort(dev sysconfig.Device, tr config.ConfGetter, opts *fsOnlyContext) error {
+	// see if anything needs to happen
+	var pristineSSHPort, newSSHPort interface{}
+
+	if err := tr.GetPristine("core", sshPortOpt, &pristineSSHPort); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if err := tr.Get("core", sshPortOpt, &newSSHPort); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if pristineSSHPort == newSSHPort {
+		return nil
+	}
+
+	// ssh.port config has changed, write new config
+	root := dirs.GlobalRootDir
+	if opts != nil {
+		root = opts.RootDir
+	}
+
+	// Note: Only UC20+ supports ussing the "sshd_config.d"
+	// dir. Supporting older systems would be hard because we
+	// would have to merge somehow the UC16 and UC18 config in a
+	// fsOnlyContext
+	if !dev.HasModeenv() {
```

Comment:
For someone like me who doesn't know what modeenv is, this conditions looks a bit hacky: it seems we are testing an unrelated feature (modeenv) because we know that this is only supported in UC20+. But if this is so, this looks almost accidental (what if we later change something and HasModeenv is not anymore synonym with "UC20+"?).

But if indeed the check is correct, I think that the comment should be expanded a bit to explain how modeenv comes into the picture.

Reply:
That is a good point. It is hacky. I need to think a bit, one issue is that at image-build time there is no easy way to check the target system. OTOH I could just omit the check and we simply document that this won't work on UC18 or UC16.